pkg/controller/linstorcsidriver/linstorcsidriver_controller.go

```diff
@@ -340,18 +345,181 @@ func (r *ReconcileLinstorCSIDriver) reconcileStatus(ctx context.Context, csiReso
+	}
+
+	return nil
+}
+
+// reconcileCSINodeObjects ensures that existing CSINode object report the right topology keys.
+//
+// Topology keys are only queried once at start-up. LINSTOR's keys are updated periodically by the operator, and so
+// the set of supported keys can change. The only reliable way to update them is restart the whole pod.
+func (r *ReconcileLinstorCSIDriver) reconcileCSINodeObjects(ctx context.Context, csiResource *piraeusv1.LinstorCSIDriver) error {
+	logger := logrus.WithFields(logrus.Fields{
+		"Name":      csiResource.Name,
+		"Namespace": csiResource.Namespace,
+		"Op":        "reconcileCSINodeObjects",
+	})
+	logger.Debug("creating linstor client")
+
+	lclient, err := lc.NewHighLevelLinstorClientFromConfig(
+		csiResource.Spec.ControllerEndpoint,
+		&csiResource.Spec.LinstorClientConfig,
+		lc.NamedSecret(ctx, r.client, csiResource.Namespace),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create linstor client: %w", err)
+	}
+
+	if !lclient.ControllerReachable(ctx) {
+		logger.Debug("controller not online, nothing to reconcile")
+
+		return nil
+	}
+
+	logger.Debug("fetching linstor nodes")
+
+	lnodes, err := lclient.Nodes.GetAll(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch lisntor nodes: %w", err)
+	}
+
+	nodePods := corev1.PodList{}
+	meta := getObjectMeta(csiResource, NodeDaemonSet, kubeSpec.CSINodeRole)
+
+	err = r.client.List(ctx, &nodePods, client.MatchingLabels(meta.Labels))
+	if err != nil {
+		return fmt.Errorf("failed to list csi node pods: %w", err)
+	}
+
+	csiNodes := storagev1.CSINodeList{}
+
+	err = r.client.List(ctx, &csiNodes)
+	if err != nil {
+		return fmt.Errorf("failed to list csi node objects: %w", err)
+	}
+
+	for i := range nodePods.Items {
+		pod := &nodePods.Items[i]
+
+		err := r.reconcileCSINodeForPod(ctx, pod, lnodes, csiNodes.Items)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcileLinstorCSIDriver) reconcileCSINodeForPod(ctx context.Context, pod *corev1.Pod, lnodes []lapi.Node, csiNodes []storagev1.CSINode) error {
+	logger := logrus.WithField("pod", pod.Name)
+
+	logger.Debug("searching matching linstor node")
+
+	lnode := nodeByName(lnodes, pod.Spec.NodeName)
+	if lnode == nil {
+		logger.Debug("no linstor node found, skipping")
+
+		return nil
+	}
+
+	logger.Debug("searching matching csi driver spec")
+
+	csiDriver := csiDriverForNode(csiNodes, pod.Spec.NodeName)
+	if csiDriver == nil {
+		logger.Debug("no csi driver found, skipping")
+
+		return nil
+	}
+
+	hasAllKeys := true
+
+	for k := range lnode.Props {
+		if !strings.HasPrefix(k, lapiconst.NamespcAuxiliary+"/") {
+			// Only Aux/ properties are used for scheduling
+			continue
+		}
+
+		expectedKey := k[len(lapiconst.NamespcAuxiliary+"/"):]
+
+		if !mdutil.SliceContains(csiDriver.TopologyKeys, expectedKey) {
+			logger.WithField("topologyKey", expectedKey).Debug("key missing in exported topology keys")
+
+			hasAllKeys = false
+
+			break
+		}
+	}
+
+	if hasAllKeys {
```

Comment:
This looks like it only checks that all linstor node properties with the matching keys are also present in `csiDriver.TopologyKeys`. So it will trigger an update if a new linstor node property is added. What happens if a property is removed or the value changed?

Reply:
It's complicated: the `csiDriver.TopologyKeys` is only a list of keys. This list is created by the kubelet when requesting the topology labels on startup. the actual values are directly passed to the k8s node. the `csiDriver.TopologyKeys` is then used by the csi-provisioner to fetch the actual labels from the k8s node.

So for removed labels: they are not present on the k8s node, so we should not have to do anything. For changed values: same idea, the k8s nodes labels are always the source of truth for the value.

Reply:
Ok. Yes, that's complicated. I'm fine with how it is. No need to make allowances for possible future behavior changes.